### PR TITLE
Fix gRPC-tunnel targets never reaching gNMIc, and Prometheus output Service port drift

### LINKED ISSUE #102: Seems like tunnel Cluster not works
I'm trying to set up tunnel cluster and have errors at one pf the cluster pods:
```
2026-07-03T06:32:21+02:00 time=2026-07-03T04:32:21.650Z level=INFO msg="tunnel server target register request" component=tunnel-server target="{ID:nbg1-dc8-vc Type:GNMI_GNOI}"
2026-07-03T06:32:21+02:00 time=2026-07-03T04:32:21.650Z level=INFO msg="target ignored, not matching any rule" component=tunnel-server target="{ID:nbg1-dc8-vc Type:GNMI_GNOI}"
2026-07-03T06:37:21+02:00 client 172.20.8.124:39592 is not in subscription list
2026-07-03T06:37:21+02:00 time=2026-07-03T04:37:21.652Z level=INFO msg="tunnel server target deregister request" component=tunnel-server target="{ID:nbg1-dc8-vc Type:GNMI_GNOI}"
2026-07-03T06:37:21+02:00 time=2026-07-03T04:37:21.815Z level=INFO msg="tunnel server target register request" component=tunnel-server target="{ID:nbg1-dc8-vc Type:GNMI_GNOI}"
2026-07-03T06:37:21+02:00 time=2026-07-03T04:37:21.815Z level=INFO msg="target ignored, not matching any rule" component=tunnel-server target="{ID:nbg1-dc8-vc Type:GNMI_GNOI}"
2026-07-03T06:42:21+02:00 time=2026-07-03T04:42:21.816Z level=INFO msg="tunnel server target deregister request" component=tunnel-server target="{ID:nbg1-dc8-vc Type:GNMI_GNOI}"
2026-07-03T06:42:21+02:00 client 172.20.8.124:39592 is not in subscription list
2026-07-03T06:42:21+02:00 time=2026-07-03T04:42:21.959Z level=INFO msg="tunnel server target register request" component=tunnel-server target="{ID:nbg1-dc8-vc Type:GNMI_GNOI}"
2026-07-03T06:42:21+02:00 time=2026-07-03T04:42:21.959Z level=INFO msg="target ignored, not matching any rule" component=tunnel-server target="{ID:nbg1-dc8-vc Type:GNMI_GNOI}"
```
I use `ghcr.io/openconfig/gnmic:0.46.0` and operator of 0.3.0

P.S.: Proposer fix in #106

### PR DESCRIPTION
**Pull request by zentavr**
Fixes #102

## Summary
A TunnelTargetPolicy-based tunnel deployment (no static Targets) never worked: the
switch/router's tunnel registration was always rejected, and even after fixing that,
the Prometheus output's Service pointed at the wrong port. Four issues, found in order
while tracing the whole path from TunnelTargetPolicy CRD to gNMIc's running config:
1. **TunnelTargetPolicy match rules were computed but never applied.** The operator
   resolved `TunnelTargetPolicy` -> `TunnelTargetMatch` but nothing in the reconcile
   path pushed the result anywhere gNMIc could see it.
2. **Empty `.spec.match` (match-all) rendered as literal empty strings, not a wildcard.**
   The final config.yaml is rendered via `gopkg.in/yaml.v2`, which does not honor `json`
   struct tags at all - `TunnelTargetMatch` only had `json` tags, so yaml.v2 serialized
   every field regardless of zero-value, turning "match everything" into `type: ""` /
   `id: ""`, which gNMIc does not treat as a wildcard.
3. **Bulk `/api/v1/config/apply` silently drops `tunnel-target-matches`.** gNMIc's own
   `mapstructure`-based decoder for that endpoint is hyphen-blind on this field. Switched
   to POSTing each `TunnelTargetMatch` individually to
   `/api/v1/config/tunnel-target-matches`, which decodes correctly.
4. **`applyConfigToPods` skipped every pod when there were zero static Targets.**
   `DistributeTargets`/`boundedLoadRendezvousHash` returns an empty `PerPodPlans` map
   when `len(Targets) == 0` (true for any tunnel-only deployment), and the per-pod loop
   did `continue` on a missing entry -- silently skipping `sendApplyRequest` *and* the
   tunnel-target-matches push for every pod, every reconcile. This meant fixes point 1- point 3
   never actually reached a running pod. Build the same fallback shape
   `DistributeTargets` would have produced (empty `Targets`, everything else from the
   plan) so subscriptions/outputs/processors/tunnel-target-matches still reach every pod.
5. **Prometheus output Service port could disagree with gNMIc's real listen port.**
   `parseListenPortAndPath` defaulted a Prometheus Output with no explicit `listen` field
   to a fabricated port 9804, which `reconcilePrometheusServices` then treated as an
   explicit user override for the generated Service (and its `prometheus.io/port`
   annotation). The port gNMIc actually binds to is computed independently and
   unconditionally by `assignPrometheusOutputPorts` -- a hash-based assignment across a
   1009-slot pool starting at 9804, so that multiple Prometheus outputs sharing one
   gNMIc process never collide. Whenever the hash didn't happen to land on 9804 itself,
   the Service and the container disagreed and scraping silently failed. Return port 0
   when `listen` isn't set so the caller falls back to the port the plan actually
   assigned.
## Testing
- Reproduced end-to-end against a real Juniper QFX5110 Virtual Chassis dialing in over
  gRPC tunnel; confirmed via operator + gNMIc pod logs that registration went from
  `"target ignored, not matching any rule"` (every ~5 min, indefinitely) to a successful
  match and running subscriptions/outputs.
- Added/updated unit tests (`helpers_test.go`) covering the Prometheus listen-port
  parsing edge cases (config present with no `listen` field, and `{}`).
- `go build ./...` and `go vet ./...` clean.

### CODE REVIEW

**Comment by karimra** [comment]
The prometheus port drift is nice catch, thanks for that.

But I think the operator should not configure the tunnel target matches statically using the config file. gNMIc was missing `mapstructure` tags for the tunnel-target-matches field which I added [here](https://github.com/openconfig/gnmic/pull/886).
So when using the next gNMIc release (v0.47.0) the operator will be able to dynamically add, update and remove tunnel target matches.

**Comment by zentavr** [comment]
@karimra so what's the plan? just wait for gNMIc release (v0.47.0) and that would be self-fixed?

File: internal/controller/cluster_controller.go [hunk 1]
```diff
@@ -486,6 +485,14 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
 	r.m.Unlock()
 
+	// reconcile statefulset (needs applyPlan.TunnelTargetMatches for tunnel-server.targets)
+	statefulSet, err := r.reconcileStatefulSet(ctx, &cluster, applyPlan.TunnelTargetMatches)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("reconciled cluster statefulset", "replicas", ptr.Deref(statefulSet.Spec.Replicas, 0), "image", statefulSet.Spec.Template.Spec.Containers[0].Image)
+
```
Comment (karimra) [inline internal/controller/cluster_controller.go:495]: There is no need to swap these steps order once gnmic properly accepts the tunnel-target-matches (v0.47.0).
Also no need to include `tunnel-server.targets` in the static config.

File: internal/controller/cluster_controller.go [hunk 2]
```diff
@@ -707,7 +714,23 @@ func (r *ClusterReconciler) applyConfigToPods(ctx context.Context, cluster *gnmi
...
+			// static target distribution, so build the same fallback shape
+			// DistributeTargets would have, just with no targets assigned.
+			podPlan = &gnmic.ApplyPlan{
+				Targets:             make(map[string]*gapi.TargetConfig),
+				Subscriptions:       plan.Subscriptions,
+				Outputs:             plan.Outputs,
+				Inputs:              plan.Inputs,
+				Processors:          plan.Processors,
+				TunnelTargetMatches: plan.TunnelTargetMatches,
+			}
```
Comment (karimra) [inline internal/controller/cluster_controller.go:733]: This is the main (operator-side) bug that prevented tunnel targets from being applied.
Let's keep it, but it would be good to wrap it in separate function.

File: internal/controller/cluster_controller.go [hunk 3]
```diff
@@ -717,13 +740,33 @@ func (r *ClusterReconciler) applyConfigToPods(ctx context.Context, cluster *gnmi
...
+		// gnmic's pkg/collector/api/server/apply.go and
+		// decodeRequestMap()). The per-resource endpoint decodes the
+		// body directly into config.TunnelTargetMatch, whose own fields
+		// (type/id/config) DO have mapstructure tags and aren't
+		// hyphenated, so it works correctly.
+		if err := r.applyTunnelTargetMatches(ctx, podBaseURL, plan.TunnelTargetMatches, httpClient); err != nil {
+			return 0, fmt.Errorf("failed to apply tunnel-target-matches to pod %d: %w", podIndex, err)
+		}
+
+		logger.Info("config applied to pod", "pod", podIndex, "targets", len(podPlan.Targets), "tunnelTargetMatches", len(plan.TunnelTargetMatches))
```
Comment (karimra) [inline internal/controller/cluster_controller.go:769]: this won't be needed anymore since tunnel-target-matches will be applied by `r.sendApplyRequest` (gnmic v0.47.0)

**Review by karimra — CHANGES_REQUESTED** [review]

**Comment by karimra** [comment]
I put comments on the parts that need to change. The goal is to only use the /config/apply endpoint from the operator so that everything gets applied at the same time, the config stays dynamic (including tunnel-targets) and is applied atomically in each pod.

**Comment by karimra** [comment]
@zentavr sorry for the slow turnaround.
Most of this is now unnecessary, through no fault of the work.

The tunnel bug is fixed upstream. gNMIc v0.47.0 (2026-08-08) added the [missing tag](https://github.com/openconfig/gnmic/blob/main/pkg/collector/api/server/apply.go#L23)

Exactly what you diagnosed. The handler also now deletes objects absent from the apply request, which covers your getTunnelTargetMatchIDs/deleteTunnelTargetMatch pair. So the per-resource POST/DELETE helpers, the reconcileStatefulSet reorder, the tunnel-server.targets rendering and the yaml tags can all go. The apply path marshals with encoding/json, so the existing `json:",omitempty"` already gives you the empty-match wildcard.

What is worth keeping is the Prometheus port drift fix. Could you open a separate PR with just that? Return 0, path, nil when listen is absent, plus your tests and one asserting an explicit listen: ":9804" is still honored.

Thanks for finding the mapstructure behavior, it was hard to see.

**Comment by karimra** [comment]
Sorry but everything here is merged or been overtaken.

- **Tunnel target matches**: fixed upstream. gNMIc v0.47.0 added the missing `mapstructure` tag, as you diagnosed, and the apply handler now also removes objects absent from the request. `/config/apply` carries `tunnel-target-matches` correctly, so the per-resource POST/DELETE helpers, the `reconcileStatefulSet` reorder, the `tunnel-server.targets` rendering and the yaml tags are all unnecessary. The apply path marshals with `encoding/json`, so the existing `json:",omitempty"` already gives the empty-match wildcard.
- **Prometheus port drift**: landed in #138. `parseListenPortAndPath` now returns `0, path, nil` when `listen` is absent, so the Service falls back to the port the plan actually assigned instead of a fabricated 9804.

Closing this out. The `mapstructure` behaviour was the hard part to see and it's what produced openconfig/gnmic#886, much appreciated.
